instance/tests/test_openstack_utils.py

```diff
@@ -292,23 +290,25 @@ def test_get_server_public_address(self, addresses, expected_address, version=4)
         server = server_class(addresses=addresses)
         self.assertEqual(openstack_utils.get_server_public_address(server, ip_version=version), expected_address)
 
-    @patch('requests.packages.urllib3.util.retry.Retry.sleep')
-    @patch('http.client.HTTPConnection.getresponse')
-    @patch('http.client.HTTPConnection.request')
-    def test_nova_client_connection_error(self, mock_request, mock_getresponse, mock_retry_sleep):
-        """
-        Connection error during a request from the nova client
-        Ensure requests are retried before giving up, with a backoff sleep between attempts
-        """
-        def getresponse_call(*args, **kwargs):
-            """ Invoked by the nova client when making a HTTP request (via requests/urllib3) """
-            raise ConnectionResetError('[Errno 104] Connection reset by peer')
-        mock_getresponse.side_effect = getresponse_call
-        nova = openstack_utils.get_nova_client(settings.OPENSTACK_REGION)
-        with self.assertRaises(keystoneauth1.exceptions.discovery.DiscoveryFailure):
```

Comment:
@clemente, it seems that for this API version we can expect `keystoneauth1.exceptions.connection.ConnectFailure` here.
cf. [session.py](https://github.com/openstack/keystoneauth/blob/master/keystoneauth1/session.py#L1012-L1020) and [test_session.py](https://github.com/openstack/keystoneauth/blob/master/keystoneauth1/tests/unit/test_session.py#L1518-L1532).

Reply:
@Agrendalath This is a more complex issue. Now I understand it, after much debugging.
- The main reason for this test failure is that I forgot to add this line to `.env.test`: `OPENSTACK_AUTH_URL_V3='http://example.com/auth/url`. It's a long story from there.
- The variable is one I introduced temporarily (I'll remove it later and we'll use the default one, `OPENSTACK_AUTH_URL`)
- The long story is: because I didn't define the _V3 in .env.test, it took the value from other place (environment, I think), and therefore `OPENSTACK_AUTH_URL_V3` ended up with the value `'https://auth.cloud.ovh.net/v3'`.
- The test doesn't run any external connection to any Internet server (I verified this), so it doesn't matter whether it says `auth.cloud.ovh.net` or `example.com`, both are fake. However, the `/v3` in `'https://auth.cloud.ovh.net/v3'` matters a lot, because keystone [is looking for it](https://github.com/openstack/keystoneauth/blob/1fd67089f7714fcaf0df1e3db845b93272b95b15/keystoneauth1/identity/generic/base.py#L155). In our original test we used `'http://example.com/auth/url'`, which doesn't contain `/v3`, and therefore it ran the last branch of the `if` and raised `DiscoveryFailure`. Because of the `/v3`, it ran another branch and caused another exception (`ConnectFailure`, it seems)
- since the test we're discussing ([test_nova_client_connection_error](https://github.com/open-craft/opencraft/blob/2b53d2049088ec7accfc408c88a605ac262141a3/instance/tests/test_openstack_utils.py#L298]) isn't testing any particular version (v2 nor v3), but just the side effects, I think it's better to keep it API-version-agnostic, and avoid mentioning any `/v2.0` or `/v3`, and leave it with `OPENSTACK_AUTH_URL='http://example.com/auth/url'` as it was before
- therefore: after I delete the `OPENSTACK_AUTH_URL_V3` and we use `OPENSTACK_AUTH_URL`, which is `'http://example.com/auth/url'`, this test won't fail. Tested in CircleCI through SSH